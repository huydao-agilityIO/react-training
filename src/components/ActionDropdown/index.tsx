import { memo } from 'react';
import {
  IconButton,
  List,
  ListItem,
  Popover,
  PopoverBody,
  PopoverContent,
  PopoverTrigger,
  Stack,
  useDisclosure
} from '@chakra-ui/react';

// Svg
import { EllipsisVertical } from '@shared/SVG';

interface Action {
  label: string;
  onClick?: () => void;
}

interface ActionDropdownProps {
  actions: Action[];
}

const ActionDropdown = ({ actions }: ActionDropdownProps) => {
  const { isOpen, onOpen, onClose } = useDisclosure();

  return (
    <Popover isOpen={isOpen} onClose={onClose} placement="bottom-start">
      <PopoverTrigger>
        <IconButton
          aria-label="show-dropdown"
          variant="default"
          color="secondary.150"
          _hover={{
            color: 'secondary.600'
          }}
          icon={<EllipsisVertical color="secondary.150" />}
          onClick={onOpen}
        />
      </PopoverTrigger>
      <PopoverContent w={{ base: 20, md: 40 }} paddingTop={3} paddingBottom={5}>
        <PopoverBody>
          <Stack>
            <List>
              {actions?.map(({ label, onClick }) => (
                <ListItem
                  fontSize="xs"
                  fontWeight="light"
                  color="secondary.200"
                  px={5}
                  py={1}
                  _hover={{
                    cursor: 'pointer',
                    color: 'primary.300'
                  }}
                  key={label}
                  onClick={onClick}>
                  {label}
                </ListItem>
              ))}
            </List>
          </Stack>
        </PopoverBody>
      </PopoverContent>
    </Popover>
  );
};

export default memo(ActionDropdown);
