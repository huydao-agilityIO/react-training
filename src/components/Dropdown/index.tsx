import { MouseEventHandler, memo } from 'react';
import {
  Avatar,
  Button,
  Heading,
  IconButton,
  Popover,
  PopoverArrow,
  PopoverBody,
  PopoverContent,
  PopoverHeader,
  PopoverTrigger,
  Text,
  VStack
} from '@chakra-ui/react';

// Svg
import { LogoutIcon } from '@shared/SVG';

interface DropdownProps {
  fullName?: string;
  major?: string;
  isOpen?: boolean;
  onMouseEnter?: MouseEventHandler<HTMLButtonElement>;
  onMouseLeave?: MouseEventHandler<HTMLButtonElement>;
}

const Dropdown = ({
  fullName = '',
  major = 'management',
  isOpen = false,
  onMouseEnter,
  onMouseLeave
}: DropdownProps) => {
  return (
    <Popover isOpen={isOpen}>
      <PopoverTrigger>
        <IconButton
          aria-label="show-dropdown"
          variant="default"
          icon={<Avatar size={{ base: 'sm', md: 'lg' }} />}
          onMouseEnter={onMouseEnter}
          onMouseLeave={onMouseLeave}
        />
      </PopoverTrigger>
      <PopoverContent
        p={7.5}
        w={52}
        bg="secondary.400"
        marginTop={{ sm: 0, md: 3 }}>
        <PopoverArrow bg="secondary.400" />
        <PopoverHeader>
          <VStack alignItems="flex-end" marginBottom={2}>
            <Text
              variant="secondary"
              size="xs"
              fontWeight="light"
              lineHeight="5"
              textTransform="capitalize">
              {major}
            </Text>
            <Heading
              color="light.300"
              fontSize="xs"
              fontWeight="medium"
              lineHeight="base"
              textTransform="capitalize">
              {fullName}
            </Heading>
          </VStack>
        </PopoverHeader>
        <PopoverBody>
          <VStack alignItems="flex-end">
            <Button
              variant="default"
              paddingX={0}
              color="light.300"
              rightIcon={<LogoutIcon />}>
              Log out
            </Button>
          </VStack>
        </PopoverBody>
      </PopoverContent>
    </Popover>
  );
};

export default memo(Dropdown);
