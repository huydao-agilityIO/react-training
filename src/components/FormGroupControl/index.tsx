import { ChangeEventHandler } from 'react';
import { FormControl, FormErrorMessage, Input } from '@chakra-ui/react';

interface FormGroupControlProps {
  type: string;
  id?: string;
  placeholder?: string;
  value?: string | number;
  errorMessage?: string;
  onChange?: ChangeEventHandler<HTMLInputElement>;
}

const FormGroupControl = ({
  type,
  id,
  placeholder,
  value,
  errorMessage,
  onChange
}: FormGroupControlProps) => (
  <FormControl isInvalid={!!errorMessage}>
    <Input
      type={type}
      id={id}
      placeholder={placeholder}
      value={value}
      variant={errorMessage ? 'error' : 'default'}
      onChange={onChange}
    />
    {errorMessage && <FormErrorMessage>{errorMessage}</FormErrorMessage>}
  </FormControl>
);

export default FormGroupControl;
