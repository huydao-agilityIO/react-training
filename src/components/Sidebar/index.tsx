import { memo } from 'react';
import { Link as ReactRouterLink, useNavigate } from 'react-router-dom';
import {
  Box,
  Container,
  List,
  ListItem,
  Stack,
  Image,
  Heading,
  HStack,
  Link
} from '@chakra-ui/react';

// Types
import { SidebarMapping } from '@shared/types';

// Svg
import HomeIcon from '@shared/SVG/HomeIcon';
import { ROUTE } from '@shared/constants';

interface SidebarProps {
  sidebarMapping: SidebarMapping[];
  urlBrand?: string;
}

const Sidebar = ({ sidebarMapping, urlBrand = '' }: SidebarProps) => {
  const navigate = useNavigate();

  const handleNavigateDashboard = () => navigate(`/${ROUTE.DASHBOARD}`);

  return (
    <Stack>
      <Container
        position="fixed"
        zIndex="dropdown"
        overflowY="auto"
        height="100vh"
        bg="light.300"
        top={0}
        left={0}
        p={0}
        width={{ lg: 52, xl: 270 }}>
        <Box px={14} py={16}>
          <Link as={ReactRouterLink} to={ROUTE.DASHBOARD}>
            <Image
              src={urlBrand}
              alt="urlBrand"
              onClick={handleNavigateDashboard}
            />
          </Link>
        </Box>
        <List>
          {sidebarMapping.map(({ heading, contentLink, pathRouter }, index) => (
            <ListItem key={index} paddingLeft={7.5}>
              <Heading
                size="lg"
                fontWeight="bold"
                paddingBottom={1}
                paddingTop={4}>
                {heading}
              </Heading>
              <HStack gap={4} py={4}>
                <HomeIcon />
                <Link
                  color="secondary.250"
                  _hover={{
                    color: 'primary.200'
                  }}
                  _focus={{
                    color: 'primary.200'
                  }}
                  as={ReactRouterLink}
                  to={pathRouter}>
                  {contentLink}
                </Link>
              </HStack>
            </ListItem>
          ))}
        </List>
      </Container>
    </Stack>
  );
};

export default memo(Sidebar);
