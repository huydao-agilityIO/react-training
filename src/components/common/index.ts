export { default as Card } from './Card';
export { default as DataTable } from './Table';
export { default as ProgressBar } from './ProgressBar';
export * from './Text';
