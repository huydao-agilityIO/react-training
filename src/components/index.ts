// Common components
export * from './common';

export * from './Author';

export { default as ActionDropdown } from './ActionDropdown';
export { default as Dropdown } from './Dropdown';
export { default as FormGroupControl } from './FormGroupControl';
export { default as Header } from './Header';
export { default as SearchBar } from './SearchBar';
export { default as Sidebar } from './Sidebar';
