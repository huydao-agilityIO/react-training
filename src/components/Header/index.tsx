import {
  Drawer,
  DrawerBody,
  DrawerCloseButton,
  DrawerContent,
  DrawerOverlay,
  HStack,
  IconButton,
  Show,
  VStack,
  useDisclosure
} from '@chakra-ui/react';

// Images
import urlBrand from '@shared/assets/images/brand.png';

// Themes
import { breakpoints } from '@shared/themes/bases';

// Constants
import { SIDEBAR_MAPPING } from '@shared/constants';

// Svg
import {
  MenuIcon,
  MessageIcon,
  NotificationIcon,
  SearchIcon
} from '@shared/SVG';

// Components
import { Dropdown, SearchBar, Sidebar } from '@shared/components';

interface HeaderProps {
  fullName?: string;
  onSearch: (value: string) => void;
}

const Header = ({ fullName, onSearch }: HeaderProps) => {
  const { md, lg } = breakpoints || {};
  const {
    isOpen: isOpenDropdown,
    onOpen: onOpenDropdown,
    onClose: onCloseDropdown
  } = useDisclosure() || {};
  const {
    isOpen: isOpenSideBar,
    onOpen: onOpenSideBar,
    onClose: onCloseSideBar
  } = useDisclosure() || {};

  return (
    <VStack p={{ base: 5, md: 7.5 }} bg="light.300">
      <HStack w="full" justifyContent="space-between">
        <HStack>
          <Show below={`${lg}px`}>
            <IconButton
              variant="default"
              aria-label="show-sidebar"
              icon={<MenuIcon />}
              onClick={onOpenSideBar}
            />
            <Drawer
              isOpen={isOpenSideBar}
              placement="left"
              onClose={onCloseSideBar}>
              <DrawerOverlay />
              <DrawerContent>
                <DrawerCloseButton />
                <DrawerBody>
                  <Sidebar
                    sidebarMapping={SIDEBAR_MAPPING}
                    urlBrand={urlBrand}
                  />
                </DrawerBody>
              </DrawerContent>
            </Drawer>
          </Show>
          <Show above={`${md}px`}>
            <SearchBar
              placeholder="Search here..."
              leftContent={<SearchIcon />}
              onChange={onSearch}
            />
          </Show>
        </HStack>
        <HStack spacing={{ base: 2, md: 7.5 }}>
          <NotificationIcon />
          <MessageIcon />
          <Dropdown
            fullName={fullName}
            isOpen={isOpenDropdown}
            onMouseEnter={onOpenDropdown}
            onMouseLeave={onCloseDropdown}
          />
        </HStack>
      </HStack>
      <Show below={`${md}px`}>
        <SearchBar
          placeholder="Search here..."
          leftContent={<SearchIcon />}
          onChange={onSearch}
        />
      </Show>
    </VStack>
  );
};

export default Header;
