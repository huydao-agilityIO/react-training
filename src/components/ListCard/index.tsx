import { List, ListItem } from '@chakra-ui/react';

// Types
import { HospitalStaff } from '@shared/types';

// Utils
import { formatFullName } from '@shared/utils';

// Components
import { Card } from '@shared/components';

interface ListCardProps {
  cards: HospitalStaff[];
}

const ListCard = ({ cards }: ListCardProps) => (
  <List display="flex" gap={7.5}>
    {cards.map(({ id, urlAvatar, firstName, lastName, description }) => (
      <ListItem key={id}>
        <Card
          url={urlAvatar}
          description={description}
          name={formatFullName(firstName, lastName)}
        />
      </ListItem>
    ))}
  </List>
);

export default ListCard;
