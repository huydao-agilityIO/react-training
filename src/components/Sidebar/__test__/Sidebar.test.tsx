import { BrowserRouter } from 'react-router-dom';

// Constants
import { SIDEBAR_MAPPING } from '@shared/constants';

// Components
import { Sidebar } from '@shared/components';

describe('Sidebar component', () => {
  const RenderContainerSidebar = () => (
    <BrowserRouter>
      <Sidebar sidebarMapping={SIDEBAR_MAPPING} />
    </BrowserRouter>
  );

  it('Match snapshot for sidebar', () => {
    const { container } = testLibReactUtils.render(<RenderContainerSidebar />);

    expect(container).toMatchSnapshot();
  });
});
