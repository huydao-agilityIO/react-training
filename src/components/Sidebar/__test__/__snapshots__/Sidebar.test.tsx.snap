// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar component Match snapshot for sidebar 1`] = `
<div>
  <div
    class="chakra-stack css-8g8ihq"
  >
    <div
      class="chakra-container css-174qhdx"
    >
      <div
        class="css-7i9x98"
      >
        <img
          alt="urlBrand"
          class="chakra-image css-0"
          src=""
        />
      </div>
      <ul
        class="css-155za0w"
        role="list"
      >
        <li
          class="css-e5uuyb"
        >
          <h2
            class="chakra-heading css-1p6fk1z"
          >
            Dashboard
          </h2>
          <div
            class="chakra-stack css-dllez"
          >
            <svg
              color="#a5adc6"
              height="20px"
              viewBox="0 0 24 24"
              width="20px"
            >
              <path
                d="M12 2.1 1 12h3v9h6v-7h4v7h6v-9h3L12 2.1z"
                fill="currentColor"
              />
            </svg>
            <a
              class="chakra-link css-yud0dq"
              href="/"
            >
              Dashboard
            </a>
          </div>
        </li>
      </ul>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
